### PR TITLE
Add Ior Monad Transformer

### PR DESCRIPTION
**Pull request by frroliveira**
Initial work for #1943. Happy to add docs and more methods (that EitherT has), if this seems ok

### CODE REVIEW

**Comment by kailuowang** [comment]
@frroliveira thanks very much for this PR. To give you a heads up, at the moment we are focusing on pushing across the RC1  finish line. This PR since is completely backward compatible, although useful, is not the highest on the priority list right now. But we will get back to it later.

**Comment by frroliveira** [comment]
@kailuowang No problem. Thanks for letting me know

File: core/src/main/scala/cats/data/IorT.scala [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+      case l @ Ior.Left(_) => F.pure(l.asInstanceOf[Ior[AA, D]])
+      case Ior.Right(b) => f(b).value
+      case Ior.Both(a, b) => F.map(f(b).value) {
+        case Ior.Left(a1) => Ior.Left(AA.combine(a, a1))
+        case Ior.Right(d) => Ior.Both(a, d)
+        case Ior.Both(a1, d) => Ior.Both(AA.combine(a, a1), d)
+      }
+    })
+
+  def flatMapF[AA >: A, D](f: B => F[Ior[AA, D]])(implicit F: Monad[F], AA: Semigroup[AA]): IorT[F, AA, D] =
```
[the remark this answers is not in the document]
Comment (frroliveira) [reply core/src/main/scala/cats/data/IorT.scala]: Current implementation needs it to call `flatMap`. I couldn't find a way to use it here, however `collectRight` can be changed to `F: FlatMap[F]`

File: core/src/main/scala/cats/data/IorT.scala [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+    flatMap(f andThen IorT.apply)
+
+  def subflatMap[AA >: A, D](f: B => Ior[AA, D])(implicit F: Monad[F], AA: Semigroup[AA]): IorT[F, AA, D] =
```
[the remark this answers is not in the document]
Comment (frroliveira) [reply core/src/main/scala/cats/data/IorT.scala]: This one can actually be `Functor`

**Comment by frroliveira** [comment]
Do you guys prefer docs to be merged with code or with separate PR?

**Comment by kailuowang** [comment]
@frroliveira it would be great to include docs in this PR as well.

**Comment by frroliveira** [comment]
Sorry it took me some time to get the docs.
Meanwhile, I have added an extra method `condF`. It would probably be useful in `EitherT` as well.

Regarding another PR that has been merged. Should I keep the order of the MonadError instances? Or switch them to be consistent with `EitherT`?

**Comment by frroliveira** [comment]
Ah, and I forgot. The apt package manager does not install the latest version of jekyll by default, so I updated the Contributing file

File: core/src/main/scala/cats/data/IorT.scala [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+}
+
+private[data] abstract class IorTInstances1 extends IorTInstances2 {
+  implicit def catsDataSemigroupForIorT[F[_], A, B](implicit F: Semigroup[F[Ior[A, B]]]): Semigroup[IorT[F, A, B]] =
+    new IorTSemigroup[F, A, B] { val F0: Semigroup[F[Ior[A, B]]] = F }
+
+  implicit def catsDataFoldableForIorT[F[_], A](implicit F: Foldable[F]): Foldable[IorT[F, A, ?]] =
+    new IorTFoldable[F, A] { val F0: Foldable[F] = F }
+
+  implicit def catsDataMonadErrorFForIorT[F[_], A, E](implicit FE: MonadError[F, E], A: Semigroup[A]): MonadError[IorT[F, A, ?], E] =
```
Comment (kailuowang) [inline core/src/main/scala/cats/data/IorT.scala]: yeah, +1 on switching the order of these two MonadError instance to be consistent with EitherT

File: core/src/main/scala/cats/data/IorT.scala [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+   * scala> import cats.implicits._
+   * scala> val o: Option[Int] = Some(3)
+   * scala> val n: Option[Int] = None
+   * scala> IorT.liftF(o)
+   * res0: cats.data.IorT[Option,Nothing,Int] = IorT(Some(Right(3)))
+   * scala> IorT.liftF(n)
+   * res1: cats.data.IorT[Option,Nothing,Int] = IorT(None)
+   * }}}
+   */
+  final def liftF[F[_], A, B](fb: F[B])(implicit F: Applicative[F]): IorT[F, A, B] = right(fb)
```
Comment (kailuowang) [inline core/src/main/scala/cats/data/IorT.scala]: do you mind also add `liftK` like this one https://github.com/SystemFw/cats/blob/4830e326acab64223fa642a2b39111665e590763/core/src/main/scala/cats/data/EitherT.scala#L365 ? If you don't have the time it's fine too, I can add one in a follow up PR.

Comment (frroliveira) [reply core/src/main/scala/cats/data/IorT.scala]: sure, no problem

**Pull request merged**